src/linalg/impl_linalg.rs

```diff
@@ -823,3 +824,190 @@ mod blas_tests {
         assert!(blas_column_major_2d::<f32, _>(&m));
     }
 }
+
+#[allow(dead_code)]
+fn general_kron<Sa, Sb, D, F, T>(
+    a: &ArrayBase<Sa, D>,
+    b: &ArrayBase<Sb, D>,
+    mut f: F,
+) -> Result<Array<T, D>, ErrorKind>
+where
+    T: Copy,
+    Sa: Data<Elem = T>,
+    Sb: Data<Elem = T>,
+    D: Dimension,
+    F: FnMut(T, T) -> T,
+{
+    let res_ndim = std::cmp::max(a.ndim(), b.ndim());
+
+    //Creates shapes completed by 1s to have res_ndim dims for each input array,
+    let a_shape_completed = {
+        let a_shape_iter = a.shape().iter().cloned().chain(std::iter::repeat(1));
+        let mut a_shape_completed = D::zeros(res_ndim);
+        for (a_shape_completed_elem, a_shape_completed_value) in a_shape_completed
+            .as_array_view_mut()
+            .iter_mut()
+            .zip(a_shape_iter)
+        {
+            *a_shape_completed_elem = a_shape_completed_value;
+        }
+        a_shape_completed
+    };
+
+    let b_shape_completed = {
+        let b_shape_iter = b.shape().iter().cloned().chain(std::iter::repeat(1));
+        let mut b_shape_completed = D::zeros(res_ndim);
+        for (b_shape_completed_elem, b_shape_completed_value) in b_shape_completed
+            .as_array_view_mut()
+            .iter_mut()
+            .zip(b_shape_iter)
+        {
+            *b_shape_completed_elem = b_shape_completed_value;
+        }
+        b_shape_completed
+    };
+
+    // Create result shape, checking that the multiplication doesn't overflow to guarantee safety below
+    let res_dim = {
+        let mut res_dim: D = D::zeros(res_ndim);
+        for ((res_dim_elem, &a_shape_value), &b_shape_value) in res_dim
+            .as_array_view_mut()
+            .iter_mut()
+            .zip(a_shape_completed.as_array_view())
+            .zip(b_shape_completed.as_array_view())
+        {
+            match a_shape_value.checked_mul(b_shape_value) {
+                Some(n) => *res_dim_elem = n,
+                None => return Err(ErrorKind::Overflow),
+            }
+        }
+        res_dim
+    };
+
+    // Reshape input arrays to compatible shapes
+    let a_reshape = a.view().into_shape(a_shape_completed).unwrap();
+    let b_reshape = b.view().into_shape(b_shape_completed.clone()).unwrap();
```

Comment:
into_shape is unfortunately not general enough to always succeed here :( I'm sorry if I have pushed you over to this solution with into_shape. We'll need to think about what we can do, because we know the completed shape is compatible.

Reply:
What's the issue with it? The layout constraint?

Reply:
It only supports c/f-contiguous arrays, unfortunately.

Reply:
The shape completed code can be much simpler I guess. Make views `av` and `bv` and use `.insert_axis()` until they both have the same number of axes.

Reply:
hm no, insert_axis changes the type, that's no good :(

Reply:
One can check if it's a dyn dimension - then convert to explicitly typed `ArrayD` and convert back to `Array<_, D>` later with `.into_dimensionality()`? Is that truly the best way? Not sure.